### PR TITLE
Update tokenizer with missing characters

### PR DESCRIPTION
**Pull request by mshannon-sil**
This feature allows the tokenizer to be updated with characters found in the source and target corpus that are not in the tokenizer's vocab. To use this feature, set a top level key called "update_tokenizer" to true in config.yml; otherwise, if it is set to false or the key does not exist, the default behavior for getting a tokenizer will be used.

Screenshot 2023-06-26 164718

When the user sets update_tokenizer to true, get_tokenizer() from the HuggingFaceConfig class will first load the tokenizer from the silnlp/assets folder that has precompiled character mapping rules for normalization. Then, it detects which characters from the src and trg corpus are missing from the tokenizer and adds them as tokens directly to the tokenizer.json file. It also adds tokens that consist of an underscore concatenated with one of the missing characters.

Also, when update_tokenizer is set to true, a feature is enabled that allows for conversion of slow tokenizers to fast tokenizers. If the project directory contains a sentencepiece.bpe.model file and there isn't a tokenizer_config.json file, it will load a slow tokenizer from that sentencepiece file, convert it into a fast tokenizer, and save it to the project directory as tokenizer.json and tokenizer_config.json.

Next steps involve improving the quality of the tokens added to the tokenizer; however, I think the progress so far is enough to be merged into the master branch.

- - -
This change is [Reviewable](https://reviewable.io/reviews/sillsdev/silnlp/171)

### CODE REVIEW

**Review by ddaspit — CHANGES_REQUESTED** [review]
> Reviewed 2 of 4 files at r1, 2 of 3 files at r2, 13 of 13 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/silnlp/171)* status: all files reviewed, 6 unresolved discussions (waiting on @mshannon-sil)

___
*[`poetry.toml` line 2 at r3](https://reviewable.io/reviews/sillsdev/silnlp/171#-N_70Y8PEfgOicnsGHjc:-N_70Y8PEfgOicnsGHjd:bw1o613) ([raw file](https://github.com/sillsdev/silnlp/blob/e5521e3140b12a0071b8e79edead034b80f415d4/poetry.toml#L2)):*
> ```TOML
> [virtualenvs]
> in-project = true
> ```

I would prefer that the developer can choose where the virtualenv is stored.
___
*[`silnlp/nmt/config.py` line 468 at r3](https://reviewable.io/reviews/sillsdev/silnlp/171#-N_7-ZKV0AirqOGmjrhK:-N_7-ZKV0AirqOGmjrhL:b-7yi5j6) ([raw file](https://github.com/sillsdev/silnlp/blob/0b9704557d6c2d37b86f4951d4d83ed1ad06625a/silnlp/nmt/config.py#L468)):*
> ```Python
>                 return
> 
>         self._build_vocabs()
> ```

Removing `_build_vocabs` will break the OpenNMT config. It should be possible to refactor the `_build_vocabs` implementation in `HuggingFaceConfig` so that it can still be called here instead of in the `get_tokenizer` method.
___
*[`silnlp/nmt/hugging_face_config.py` line 313 at r3](https://reviewable.io/reviews/sillsdev/silnlp/171#-N_6zm1L9ccxTZfsCBOX:-N_6zm1L9ccxTZfsCBOY:bo4gbxn) ([raw file](https://github.com/sillsdev/silnlp/blob/e5521e3140b12a0071b8e79edead034b80f415d4/silnlp/nmt/hugging_face_config.py#L313)):*
> ```Python
>                     add_lang_code_to_tokenizer(self._tokenizer, lang_code)
>                     if self.root.get("update_tokenizer"):
>                         self._add_tokens([lang_code])
> ```

It would be more efficient to only call `_add_tokens` once.
___
*[`silnlp/nmt/hugging_face_config.py` line 317 at r3](https://reviewable.io/reviews/sillsdev/silnlp/171#-N_7-6ze6oxg9qp-XORT:-N_7-6ze6oxg9qp-XORU:b-qnd9pa) ([raw file](https://github.com/sillsdev/silnlp/blob/e5521e3140b12a0071b8e79edead034b80f415d4/silnlp/nmt/hugging_face_config.py#L317)):*
> ```Python
>             if updated:
>                 self._tokenizer.save_pretrained(self.exp_dir)
>         if self.root.get("update_tokenizer"):
> ```

I think it would be better to move `update_tokenizer` under `data`.
___
*[`silnlp/nmt/hugging_face_config.py` line 322 at r3](https://reviewable.io/reviews/sillsdev/silnlp/171#-N_6v_W_EvGj5WLOqNHw:-N_6v_W_EvGj5WLOqNHx:b-k92dw8) ([raw file](https://github.com/sillsdev/silnlp/blob/e5521e3140b12a0071b8e79edead034b80f415d4/silnlp/nmt/hugging_face_config.py#L322)):*
> ```Python
>             )
>             if missing_characters:
>                 missing_characters_underscore = ["_" + c for c in missing_characters]
> ```

I don't think this is really necessary. The NLLB vocab already includes the `▁` as a token, so just including the missing character as a token should be sufficient to tokenize a text without `<unk>` tokens. Also, sentencepiece actually uses the special U+2581 character and not the underscore character.
___
*[`silnlp/nmt/hugging_face_config.py` line 337 at r3](https://reviewable.io/reviews/sillsdev/silnlp/171#-N_6uWcJCRDDsyRp_Mie:-N_6uWcJCRDDsyRp_Mif:b-bbq3wm) ([raw file](https://github.com/sillsdev/silnlp/blob/e5521e3140b12a0071b8e79edead034b80f415d4/silnlp/nmt/hugging_face_config.py#L337)):*
> ```Python
>                 self._tokenizer.save_pretrained(str(self.exp_dir))
>             else:
>                 model_name_or_path = (
> ```

This line of code is a bit complicated. It would probably be better to turn this into a normal `if` statement.

**Comment by ddaspit** [comment]
How is progress going on this?

**Comment by mshannon-sil** [comment]
Progress is going well. The updates to the tokenizer are finished outside of some style changes. However, we noticed a bit of a dip in accuracy for some languages when using new tokens, so we're working on some approaches to adapt the model to the new tokens. In particular, there's a paper by Meta AI that outlines some techniques for adding new languages. Here's the relevant issue for that: https://github.com/sillsdev/silnlp/issues/189. I'm aiming to have this done around the middle of next week, and then I'll work on merging the branches and request feedback on the code again.

**Comment by ddaspit** [comment]
When you say "dip in accuracy", what is this in comparison to? Is it a dip in accuracy compared to training the model without adding the new tokens?

**Comment by mshannon-sil** [comment]
It's in comparison to translating without updating the tokenizer, although it's also true in comparison to just adding missing characters. Some languages did see an improvement using the new trained tokens, but not many of them. For example, the drop in accuracy was a bit over 1 unit of BLEU score on the Rajbanshi New Testament.

**Comment by ddaspit** [comment]
That's interesting. Are these new tokens on the source side or target side? How many languages have you tested? This might indicate that simply adding missing characters to the vocab is not a good idea and we need a better approach.

**Comment by mshannon-sil** [comment]
The tokens are on the trg side. I've mostly been testing with Rajbanshi, and Bethany tested with four other languages, the results of which are in [this spreadsheet](https://docs.google.com/spreadsheets/d/1rBple7E2gsD0Fn-rp1VeCLDsgaAlk_oj13MQ6Fz9emw/edit#gid=0).

**Review by mshannon-sil — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sillsdev/silnlp/171)* status: 13 of 17 files reviewed, 6 unresolved discussions (waiting on @ddaspit)

___
*[`silnlp/nmt/hugging_face_config.py` line 313 at r3](https://reviewable.io/reviews/sillsdev/silnlp/171#-N_6zm1L9ccxTZfsCBOX:-NbWYDuf-KSpmyGMQgn7:b-896fix) ([raw file](https://github.com/sillsdev/silnlp/blob/e5521e3140b12a0071b8e79edead034b80f415d4/silnlp/nmt/hugging_face_config.py#L313)):*
<details><summary><i>Previously, ddaspit (Damien Daspit) wrote…</i></summary><blockquote>

It would be more efficient to only call `_add_tokens` once.
</blockquote></details>

Done.
___
*[`silnlp/nmt/hugging_face_config.py` line 317 at r3](https://reviewable.io/reviews/sillsdev/silnlp/171#-N_7-6ze6oxg9qp-XORT:-NbW_JUcEMd8T2T--xSK:b-896fix) ([raw file](https://github.com/sillsdev/silnlp/blob/e5521e3140b12a0071b8e79edead034b80f415d4/silnlp/nmt/hugging_face_config.py#L317)):*
<details><summary><i>Previously, ddaspit (Damien Daspit) wrote…</i></summary><blockquote>

I think it would be better to move `update_tokenizer` under `data`.
</blockquote></details>

Done.
___
*[`silnlp/nmt/hugging_face_config.py` line 322 at r3](https://reviewable.io/reviews/sillsdev/silnlp/171#-N_6v_W_EvGj5WLOqNHw:-NbW_L3VEqup6VHxSNl5:b-896fix) ([raw file](https://github.com/sillsdev/silnlp/blob/e5521e3140b12a0071b8e79edead034b80f415d4/silnlp/nmt/hugging_face_config.py#L322)):*
<details><summary><i>Previously, ddaspit (Damien Daspit) wrote…</i></summary><blockquote>

I don't think this is really necessary. The NLLB vocab already includes the `▁` as a token, so just including the missing character as a token should be sufficient to tokenize a text without `<unk>` tokens. Also, sentencepiece actually uses the special U+2581 character and not the underscore character.
</blockquote></details>

Done.
___
*[`silnlp/nmt/hugging_face_config.py` line 337 at r3](https://reviewable.io/reviews/sillsdev/silnlp/171#-N_6uWcJCRDDsyRp_Mie:-NbW_MS89IYDr1jAHbXw:b-896fix) ([raw file](https://github.com/sillsdev/silnlp/blob/e5521e3140b12a0071b8e79edead034b80f415d4/silnlp/nmt/hugging_face_config.py#L337)):*
<details><summary><i>Previously, ddaspit (Damien Daspit) wrote…</i></summary><blockquote>

This line of code is a bit complicated. It would probably be better to turn this into a normal `if` statement.
</blockquote></details>

Done.
___
*[`poetry.toml` line 2 at r3](https://reviewable.io/reviews/sillsdev/silnlp/171#-N_70Y8PEfgOicnsGHjc:-NbUh1s9DzKAWamGVAO0:b-896fix) ([raw file](https://github.com/sillsdev/silnlp/blob/e5521e3140b12a0071b8e79edead034b80f415d4/poetry.toml#L2)):*
<details><summary><i>Previously, ddaspit (Damien Daspit) wrote…</i></summary><blockquote>

I would prefer that the developer can choose where the virtualenv is stored.
</blockquote></details>

Done.
___
*[`silnlp/nmt/config.py` line 468 at r3](https://reviewable.io/reviews/sillsdev/silnlp/171#-N_7-ZKV0AirqOGmjrhK:-NbUh7Xo7WNmrfjXzNuf:b-896fix) ([raw file](https://github.com/sillsdev/silnlp/blob/0b9704557d6c2d37b86f4951d4d83ed1ad06625a/silnlp/nmt/config.py#L468)):*
<details><summary><i>Previously, ddaspit (Damien Daspit) wrote…</i></summary><blockquote>

Removing `_build_vocabs` will break the OpenNMT config. It should be possible to refactor the `_build_vocabs` implementation in `HuggingFaceConfig` so that it can still be called here instead of in the `get_tokenizer` method.
</blockquote></details>

Done.

**Comment by mshannon-sil** [comment]
@ddaspit I just pushed a couple commits that address the requested changes. Also, since you last reviewed the code, I've added the trained tokens as well as an option for unk embeddings. I've attached a picture of an up to date example of a config.yml file that uses the tokenizer.
Screenshot 2023-08-10 190157

**Review by ddaspit — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r4, 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/silnlp/171)* status: all files reviewed, 2 unresolved discussions (waiting on @mshannon-sil)

___
*[`silnlp/nmt/hugging_face_config.py` line 23 at r6](https://reviewable.io/reviews/sillsdev/silnlp/171#-NbprGvWCU1oHFbmurha:-NbprGvX6jVnDsVjLnGF:b-da8037) ([raw file](https://github.com/sillsdev/silnlp/blob/b447476bdf8997615d2a4190ef48b28341aef455/silnlp/nmt/hugging_face_config.py#L23)):*
> ```Python
> from torch.utils.checkpoint import checkpoint  # noqa: 401
> from transformers import (
>     AdamW,
> ```

Why was this imported?
___
*[`silnlp/nmt/hugging_face_config.py` line 29 at r6](https://reviewable.io/reviews/sillsdev/silnlp/171#-NbprKJP2DZa0PRXjcL1:-NbprKJP2DZa0PRXjcL2:b-da8037) ([raw file](https://github.com/sillsdev/silnlp/blob/b447476bdf8997615d2a4190ef48b28341aef455/silnlp/nmt/hugging_face_config.py#L29)):*
> ```Python
>     DataCollatorForSeq2Seq,
>     EarlyStoppingCallback,
>     get_constant_schedule,
> ```

Why was this imported?

**Review by ddaspit — APPROVED** [review]
:lgtm: Good Job.

> Reviewed 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/silnlp/171#-:-Nbu-Wl27lwkxcKYOv0B:b-elr8jf)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @mshannon-sil)

**Review by ddaspit — APPROVED** [review]
> Reviewed 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/silnlp/171)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @mshannon-sil)

**Comment by ddaspit** [comment]
This looks ready to go. Let's try to get this in.
